mino/minogrpc/routing.go

```diff
@@ -0,0 +1,201 @@
+package minogrpc
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/binary"
+	"fmt"
+	math "math"
+	"math/rand"
+	"sort"
+	"strings"
+
+	"go.dedis.ch/fabric"
+	"go.dedis.ch/fabric/mino"
+	"golang.org/x/xerrors"
+)
+
+// TreeRouting holds the routing tree of a network. It allows each node of the
+// tree to know which child it should contact in order to relay a message that
+// is in it sub-tree.
+type TreeRouting struct {
+	root         *treeNode
+	me           *treeNode
+	routingNodes map[string]*treeNode
+}
+
+// NewTreeRouting creates the network tree in a deterministic manner based on
+// the list of addresses. Please be careful that the provided list of addresses
+// is shuffled, which could affect subsequent use of this list, especially
+// outside the function's scope.
+func NewTreeRouting(addrs []mino.Address, addr mino.Address,
+	treeHeight int) (*TreeRouting, error) {
+
+	addrsStr := make([]string, len(addrs))
+	for i, addr := range addrs {
+		addrsStr[i] = addr.String()
```

Comment:
Use `addr.MarshalText()` instead as you don't know if the string representation is biased.

Reply:
To sort, you can define a slice type then use `bytes.Compare`.